app/routes/_gcn.circulars.edit.$circularId/CircularEditForm.tsx

```diff
@@ -0,0 +1,350 @@
+import type { Circular } from '../_gcn.circulars/circulars.lib'
+import { bodyIsValid, subjectIsValid } from '../_gcn.circulars/circulars.lib'
+import { CircularsKeywords } from '~/components/CircularsKeywords'
+import Spinner from '~/components/Spinner'
+
+function SyntaxExample({
+  label,
+  children,
+  href,
+  ...props
+}: {
+  label: ReactNode
+  children: string
+} & Pick<JSX.IntrinsicElements['a'], 'href' | 'rel'>) {
+  const { target, rel } = useContext(AstroDataContext)
+  return (
+    <tr>
+      <td>
+        {href ? (
+          <a href={href} {...props} target={target} rel={rel}>
+            {label}
+          </a>
+        ) : (
+          label
+        )}
+      </td>
+      <td>
+        <code>{children}</code>
+      </td>
+      <td>
+        <MarkdownBody>{children}</MarkdownBody>
+      </td>
+    </tr>
+  )
+}
+
+function SyntaxReference() {
+  return (
+    <>
+      <h3>GCN Circulars cross-reference syntax</h3>
+      <Table compact bordered>
+        <thead>
+          <tr>
+            <th>Service</th>
+            <th>Example</th>
+            <th>Output</th>
+          </tr>
+        </thead>
+        <tbody>
+          <SyntaxExample label="GCN Circulars" href="/circulars">
+            GCN 34653, 34660, and 34677
+          </SyntaxExample>
+          <SyntaxExample label="DOI" href="https://www.doi.org" rel="external">
+            doi:10.1103/PhysRevLett.116.061102
+          </SyntaxExample>
+          <SyntaxExample label="arXiv" href="https://arxiv.org" rel="external">
+            arXiv:1602.03837
+          </SyntaxExample>
+          <SyntaxExample
+            label="TNS"
+            href="https://www.wis-tns.org"
+            rel="external"
+          >
+            AT2017gfo
+          </SyntaxExample>
+          <SyntaxExample label="Web address">
+            https://www.swift.psu.edu
+          </SyntaxExample>
+        </tbody>
+      </Table>
+    </>
+  )
+}
+
+export function CircularEditForm({
+  formattedContributor,
+  circular,
+  defaultBody,
+  defaultSubject,
+  searchString,
+  isAuthorized,
+}: {
+  formattedContributor: string
+  circular?: Circular
+  defaultBody: string
+  defaultSubject: string
+  searchString: string
+  isAuthorized: boolean
+}) {
+  let formSearchString = '?index'
+  if (searchString) {
+    formSearchString = `${formSearchString}&${searchString}`
+    searchString = `?${searchString}`
+  }
+  const [subjectValid, setSubjectValid] = useState(
+    subjectIsValid(defaultSubject)
+  )
+  const [body, setBody] = useState(defaultBody)
+  const bodyValid = bodyIsValid(body)
+  const [showKeywords, toggleShowKeywords] = useStateToggle(false)
+  const [showBodySyntax, toggleShowBodySyntax] = useStateToggle(false)
+  const [showPreview, setShowPreview] = useState(false)
+  const sending = Boolean(useNavigation().formData)
+  const valid = subjectValid && bodyValid
+  return (
+    <>
+      <h1>{circular ? 'Edit' : 'New'} GCN Circular</h1>
```

Comment:
This PR still contains new code for history support. Would you please make sure that this PR is purely a refactoring of existing code?

Reply:
Should be all set now